From 2d4b7c0d7275d418677713461089b6f1afde2436 Mon Sep 17 00:00:00 2001
From: "Zhong,Fangjian" <fangjian.zhong@intel.com>
Date: Thu, 15 Jun 2023 22:20:16 +0800
Subject: [PATCH 07/54] fastboot: Erase stored rollback slots when status
 changes.

Stored rollback slots should be removed when status changes during
fastboot flash scenario.

Tracked-On: OAM-110761
Signed-off-by: Zhong,Fangjian <fangjian.zhong@intel.com>
---
 libfastboot/fastboot_flashing.c | 27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/libfastboot/fastboot_flashing.c b/libfastboot/fastboot_flashing.c
index a7af14f..c9ad1ae 100644
--- a/libfastboot/fastboot_flashing.c
+++ b/libfastboot/fastboot_flashing.c
@@ -40,6 +40,8 @@
 #include "gpt.h"
 #include "intel_variables.h"
 #include "android.h"
+#include "tpm2_security.h"
+#include "vars.h"
 
 static cmdlist_t cmdlist;
 
@@ -114,6 +116,31 @@ EFI_STATUS change_device_state(enum device_state new_state, BOOLEAN interactive)
 	}
 #endif
 
+	info(L"Erasing rollback index...");
+
+#ifdef USE_TPM
+/* There are only 8 rollback slots inside TPM */
+#define MAX_ROLLBACK_SLOTS_NUM 8
+#else
+#define MAX_ROLLBACK_SLOTS_NUM AVB_MAX_NUMBER_OF_ROLLBACK_INDEX_LOCATIONS
+#endif
+
+	for (int idx = 0; idx < MAX_ROLLBACK_SLOTS_NUM; idx++) {
+#ifdef USE_TPM
+		ret = write_rollback_index_tpm2(idx, 0);
+#else
+		ret = write_efi_rollback_index(idx, 0);
+#endif
+		if (EFI_ERROR(ret)) {
+			if (interactive)
+				fastboot_fail("Couldn't write rollback index");
+			else
+				info(L"Couldn't write rollback index");
+
+			return ret;
+		}
+	}
+
 	ret = set_current_state(new_state);
 	if (EFI_ERROR(ret)) {
 		if (interactive)
-- 
2.25.1

